index/segment/mem/types.go

```diff
@@ -50,15 +42,69 @@ type Options interface {
 	// InstrumentOptions returns the instrument options.
 	InstrumentOptions() instrument.Options
 
-	// SetInsertQueueSize sets the insert queue size.
-	SetInsertQueueSize(value int64) Options
+	// SetPostingsListPool sets the PostingsListPool.
+	SetPostingsListPool(value segment.PostingsListPool) Options
+
+	// PostingsListPool returns the PostingsListPool.
+	PostingsListPool() segment.PostingsListPool
+
+	// SetInitialCapacity sets the initial capacity.
+	SetInitialCapacity(value int) Options
 
-	// InsertQueueSize returns the insert queue size.
-	InsertQueueSize() int64
+	// InitialCapacity returns the initial capacity.
+	InitialCapacity() int
+}
+
+// termDictionary represents a mapping from doc.Field -> doc
+type termsDictionary interface {
+	// Insert inserts a mapping from field `f` to document ID `docID`.
+	Insert(f doc.Field, docID segment.DocID) error
+
+	// Fetch returns all the docIDs matching the given filter.
+	// NB(prateek): the returned PostingsList is safe to modify.
+	Fetch(filterName []byte, filterValue []byte, isRegex bool) (segment.PostingsList, error)
+}
 
-	// SetNumInsertWorkers sets the number of insert workers.
-	SetNumInsertWorkers(value int) Options
+// postingsManagerOffset is an offset used to reference a postingsList in the postingsManager.
+type postingsManagerOffset int
+
+// postingsManager represents an efficient mapping from postingsManagerOffset -> PostingsList
+type postingsManager interface {
+	// Insert inserts document `i` at a new postingsManagerOffset, and returns the
+	// offset used.
+	Insert(i segment.DocID) postingsManagerOffset
+
+	// Update updates postingsManagerOffset `p` to include a reference to document `i`.
+	Update(p postingsManagerOffset, i segment.DocID) error
+
+	// Fetch retrieves all documents which are associated with postingsManagerOffset `p`.
+	Fetch(p postingsManagerOffset) (segment.ImmutablePostingsList, error)
+}
+
+// predicate returns a bool indicating if the document matched the
+// provided criterion, or not.
+type predicate func(d doc.Document) bool
+
+// queryable is the base contract required for any mem segement implementation to be used by a `searcher`.
+type queryable interface {
+	// Filter retrieves the PostingsList for the filter, and any additional
+	// filtering criterion that must be applied in post processing.
+	Filter(
+		filterName []byte,
+		filterValue []byte,
+		isRegex bool,
+	) (
+		candidateDocIDs segment.PostingsList,
+		pendingFilterFn predicate,
+		err error,
+	)
+
+	// Fetch retrieves the list of documents in the given postings list.
+	Fetch(p segment.PostingsList) ([]doc.Document, error)
```

Comment:
It seems a little strange to me that this interface has methods to return both postings lists and documents. If we are guaranteeing that the set of documents contained in segments is disjoint then it seems we should just support a `Fetch` on segments method which takes in a query and returns the documents since the postings list would be a detail internal to the segment. Then the index which contains multiple segments would be responsible for merging those documents into a single result.

Reply:
These methods are used by a `searcher` which is internal to a segment, these methods can't be used by external consumers. The abstraction afforded to external users is the one you're suggesting ~ `segment.Query(q Query)`

Reply:
Oh, I see now, `Filter` gets the postings list for the individual terms, then the `searcher` does what it needs to based on the query (intersect for And, union for OR, etc.) and then retrieves the actual documents with `Fetch`?

Reply:
Yep exactly.